### PR TITLE
feat(eslint-plugin): add rule call-super-on-override

### LINKED ISSUE #684: Rule proposal: checks if you remembered to call super.method()
**Expected Result**

A rule that can detects if we forget to call a super method. Most of the time I want to call super methods. It is possible to forget to call a super method, and it can cause subtle bugs that aren't discoverable until later (depending on the implementation of the method of course).

**Actual Result**

There's not a rule that I know of for this yet.

**Additional Info**

Originally posted at https://github.com/eslint/eslint/issues/11969, but then we realized that there's no type info, so for plain JS ESLint might not be able to know if a super method exists in order to verify if you are calling super.

### PR DESCRIPTION
**Pull request by mahdi-farnia**
## PR Checklist

- [x] Addresses an existing open issue: fixes #684 
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Before:

1. Warning expected to note that overridden method should call it's super method to prevent missing something!

```ts
// Trivia example
class Foo {
  dealloc(): void {
    memory.dealloc();
  }
}

class Bar extends Foo {
  override dealloc(): void {
    console.log("Deallocated"); // note that memory.dealloc() didn't call
  }
}
```

After:

```ts
// Trivia example
class Foo {
  dealloc(): void {
    memory.dealloc();
  }
}

class Bar extends Foo {
  override dealloc(): void {
    console.log("Deallocated"); // note that memory.dealloc() didn't call
    // ^^^ missingSuperMethodCall warning
  }
}
```

> This checking can be disabled via option `ignoreMethods: true`

2. Also ordering may be important:

Before:
```ts
// Trivia example on drawing a circle with Canvas API
class MyCircle extends Circle {
  override draw(): void {
    this.changeColor("blue");
    super.draw(); // <-- ordering matters here
  }
}
```

After:
```ts
// Trivia example on drawing a circle with Canvas API
class MyCircle extends Circle {
  override draw(): void {
    this.changeColor("blue");
    super.draw();
    // ^^^ topLevel warning
  }
}
```

> This options is disabled by default and can be enabled by `topLevel: true`

Closes #684

> Note about local git history by me: I rewrite it again after one month, and i figured out that the last clone is outdated and i cloned the latest repo and just copied & pasted the file.
Because of that there is no history available for this project.

### CODE REVIEW

File: packages/eslint-plugin/src/rules/call-super-on-override.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+  },
+  defaultOptions: [
+    {
+      ignoreMethods: false,
+      topLevel: false,
+    },
+  ],
+  create(context, [{ ignoreMethods, topLevel }]) {
+    return {
+      'MethodDefinition[override=true][kind="method"]'(
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/call-super-on-override.ts:53]: Nice use of ESQuery to filter on the method definition! Very clean. 😁

File: packages/eslint-plugin/src/rules/call-super-on-override.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+      description:
+        'Require overridden methods to call super.method in their body',
+      recommended: 'warn',
+      requiresTypeChecking: false,
+    },
+    messages: {
+      missingSuperMethodCall:
+        "Use 'super.{{property}}{{parameterTuple}}' to avoid missing super class method implementations",
+      topLevelSuperMethodCall:
+        "super method must be called before accessing 'this' in the overridden method of a derived class.",
```
[the remark this answers is not in the document]
Comment (mahdi-farnia) [reply packages/eslint-plugin/src/rules/call-super-on-override.ts:27]: I don't want to impose my opinion, but in my opinion this option can be suitable for methods (not just constructors).
I leave it for now to here your opinion again & then I'll do whatever is suitable. (remove or keep...)

File: packages/eslint-plugin/src/rules/call-super-on-override.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+      ignoreMethods: false,
+      topLevel: false,
+    },
+  ],
+  create(context, [{ ignoreMethods, topLevel }]) {
+    return {
+      'MethodDefinition[override=true][kind="method"]'(
+        node: TSESTree.MethodDefinition,
+      ): void {
+        if (ignoreMethods) {
```
[the remark this answers is not in the document]
Comment (mahdi-farnia) [reply packages/eslint-plugin/src/rules/call-super-on-override.ts]: 😅

File: packages/eslint-plugin/src/rules/call-super-on-override.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+              statement,
+              methodName,
+              !methodNameIsLiteral && isComputed,
+            )
+          ) {
+            return; // We are done here, no missingSuperMethodCall error
+          }
+        }
+
+        // Raise if not found
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/call-super-on-override.ts:71]: [Nit] This is self-evident from the code, I think.

```suggestion
```

File: packages/eslint-plugin/src/rules/call-super-on-override.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        // Search for super method call
+        for (const statement of bodyStatements) {
+          if (
+            isSuperMethodCall(
+              statement,
+              methodName,
+              !methodNameIsLiteral && isComputed,
+            )
+          ) {
+            return; // We are done here, no missingSuperMethodCall error
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/call-super-on-override.ts:67]: [Nit] This is self-evident from the code, I think.

```suggestion
            return;
```

File: packages/eslint-plugin/src/rules/call-super-on-override.ts [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        node: TSESTree.MethodDefinition,
+      ): void {
+        let methodName = '',
+          methodNameIsLiteral = false,
+          methodNameIsNull = false; // don't add quotes for error message on [null] case
+
+        if (node.key.type === AST_NODE_TYPES.Identifier) {
+          methodName = node.key.name;
+        } else {
+          methodNameIsLiteral = true;
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/call-super-on-override.ts:48]: [Refactor] Aha, a duplication of information! `methodNameIsLiteral` will always be `true` if `methodNameIsNull` is `true`. So this is using two variables to store a single piece of info. I'd suggest using either a union type or enum to consolidate down to one variable.

For example, the enum form might look something like this (but maybe with different names, I haven't thought too hard about these):

```ts
type MethodNameVariant = 'identifier' | 'non-null-literal' | 'null-literal';
```

File: packages/eslint-plugin/src/rules/call-super-on-override.ts [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+  },
+  defaultOptions: [],
+  create(context) {
+    return {
+      'MethodDefinition[override=true][kind="method"]'(
+        node: TSESTree.MethodDefinition,
+      ): void {
+        let methodName = '',
+          methodNameIsLiteral = false,
+          methodNameIsNull = false; // don't add quotes for error message on [null] case
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/call-super-on-override.ts:43]: [Refactor] This code creates a collection of `let` variables once and then never changing their values again. It feels to me like that's an opportunity to  extract into a separate function to make it more clear what's going on with these. Either that, or using a ` = ? ... : ...` ternary inline. What do you think?

Combining with the deduplication suggestion (and again not thinking too hard on the name):

```ts
const [methodName, methodVariant] = determineMethodNameAndVariant(node);
```

File: packages/eslint-plugin/src/rules/call-super-on-override.ts [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+export default utils.createRule<[], MessageIds>({
+  name: 'call-super-on-override',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        'Require overridden methods to call super.method in their body',
+      recommended: false,
+      requiresTypeChecking: false,
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/call-super-on-override.ts:16]: It's fine to leave this out; falsy is its default.

```suggestion
```

File: packages/eslint-plugin/src/rules/call-super-on-override.ts [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+          methodNameIsNull = false; // don't add quotes for error message on [null] case
+
+        if (node.key.type === AST_NODE_TYPES.Identifier) {
+          methodName = node.key.name;
+        } else {
+          methodNameIsLiteral = true;
+          // null & undefined can be used as property names, undefined counted as Identifier & null as Literal
+          methodName =
+            (node.key as TSESTree.Literal).value?.toString() ?? 'null';
+          methodNameIsNull = (node.key as TSESTree.Literal).value == null;
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/call-super-on-override.ts:52]: > `node.key as TSESTree.Literal`

[Refactor] This `as` assertion isn't always going to be correct. The TSESTree types are set up to represent all the possible values of properties. The fact that an `as` was needed to get this code to type check means that the code isn't handling some case that it needs to handle.

Removing the `as` gives:

```plaintext
Property 'value' does not exist on type 'FunctionExpression | ArrayExpression | ArrayPattern | ArrowFunctionExpression | AssignmentExpression | ... 35 more ... | YieldExpression'.
  Property 'value' does not exist on type 'FunctionExpression'.
```

...which is true. [What if the node's key is a function expression](
https://typescript-eslint.io/play/#ts=4.9.3&sourceType=module&showAST=es&code=MYGwhgzhAECSB2A3MICWATAymAtgBxAFNoBvAKGmgG0AzAV3mABdUB7eaACgEpToBfALo9SFStGDsIrIgDoQrAOacARAE1UePIUIBCFdzH8y-IA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)?

```ts
class InvalidSample {
  [function () { }]() {
    console.log("Yippee!")
  }
}
```

This code is wacky (and causes a TypeScript complaint) but someone in the wild might try to run this rule on that code. 

Change request: remove all the `as` assertions from this file, and explicitly handle those odd node edge cases. 

Note too that we require high unit test coverage (generally, ~100% if possible) for new rule tests. So this'll need unit tests added for any new logic. You can always run `yarn test` from `packages/eslint-plugin` to generate a testing coverage report at `packages/eslint-plugin/coverage/lcov-report/index.html` (or see the Codecov PR comment for an online viewer).

File: packages/eslint-plugin/docs/rules/call-super-on-override.md [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+class Foo2 extends Foo1 {
+  override bar(param: any): void {
+    super.bar(param);
+  }
+}
+```
+
+## When Not To Use It
+
+When you are using TypeScript < 4.3 or you did not set `noImplicitOverride: true` in `CompilerOptions`
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/docs/rules/call-super-on-override.md:43]: A lot of consumers don't use `noImplicitOverride`. I think we'll still want the rule to be able to work for them... which means it'll have to use [rule type checking](https://typescript-eslint.io/custom-rules#type-checking) when there isn't an `override` operator on a method.

Roughly, that means:

1. Grabbing the type of the extended class
2. Checking whether it has a method / function property under the same name

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Getting closer, thanks for continuing to work on this! Unless someone in @typescript-eslint/triage-team disagrees with me, I think we'll want type checking in the rule?

File: packages/eslint-plugin/docs/rules/call-super-on-override.md [hunk 11]
```diff
@@ -0,0 +1,43 @@
+---
+description: 'Require overridden methods to call super.method in their body.'
+---
+
+> 🛑 This file is source code, not the primary documentation location! 🛑
+>
+> See **https://typescript-eslint.io/rules/call-super-on-override** for documentation.
+
+This rule enforces that overridden methods are calling exact super method to avoid missing super class method implementations.
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/docs/rules/call-super-on-override.md:9]: [Docs] A little more clear, I think?

```suggestion
This rule enforces that overridden methods call their corresponding `super` method.
Doing so can be useful in architectures where base class methods are meant to always be called by child classes.
```

**Comment by JoshuaKGoldberg** [comment]
👋 @mahdi-farnia! Just checking in, is this still something you have time for? No worries if not - I just don't want to leave it hanging.

**Comment by mahdi-farnia** [comment]
I'll name this laziness of myself 

Sorry about that
I am going to check it tonight ( My country has 12:30 hours difference from US )

I forgot resolving issues after my college exams...🤦‍♂️

**Comment by mahdi-farnia** [comment]
I need help!
How can grabbing the type of the extended class(es)?
Any article or guid?
@JoshuaKGoldberg

**Comment by mahdi-farnia** [comment]
Just note that i did not make any progress (for now).
I just applied some suggested changes

I already using [astexplorer](http://astexplorer.net) and [ts-ast-viewer](http://ts-ast-viewer.com/)

Is there any other tool?

File: packages/eslint-plugin/docs/rules/call-super-on-override.md (hunk 10 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/docs/rules/call-super-on-override.md]: Let's keep conversations threaded, please, so it's easier to have multiple at a time & mark them as open or resolved.

As to _how_ to grab the type of the extended class: you'll need to use the type checker. Have you read our resources on doing that?
* https://typescript-eslint.io/blog/asts-and-typescript-eslint
* https://sourcegraph.com/notebooks/Tm90ZWJvb2s6MTA2OA==
* https://typescript-eslint.io/custom-rules#typed-rules

**Comment by JoshuaKGoldberg** [comment]
I responded in this thread: https://github.com/typescript-eslint/typescript-eslint/pull/5848#discussion_r1092552144
